library/ssl_tls.c

```diff
@@ -6854,13 +6855,33 @@ int mbedtls_ssl_get_record_expansion( const mbedtls_ssl_context *ssl )
     {
         case MBEDTLS_MODE_GCM:
         case MBEDTLS_MODE_CCM:
+        case MBEDTLS_MODE_CHACHAPOLY:
         case MBEDTLS_MODE_STREAM:
             transform_expansion = transform->minlen;
             break;
 
         case MBEDTLS_MODE_CBC:
-            transform_expansion = transform->maclen
-                      + mbedtls_cipher_get_block_size( &transform->cipher_ctx_enc );
+
+            block_size = mbedtls_cipher_get_block_size(
+                &transform->cipher_ctx_enc );
+
+#if defined(MBEDTLS_SSL_PROTO_TLS1_1) || defined(MBEDTLS_SSL_PROTO_TLS1_2)
+            if( ssl->minor_ver >= MBEDTLS_SSL_MINOR_VERSION_2 )
+            {
+                /* Expansion due to addition of
```

Comment:
This logic is largely duplicated in `mbedtls_ssl_derive_keys` with the calculation of `transform->minlen`. We have a mostly-duplicated calculation, and we got it wrong in one place, so we really should de-duplicate it.

`transform->minlen` is the size of a record with a payload of size 0, i.e. it's the record overhead for a payload of size 0. `get_record_expansion` is the maximum record overhead. The difference between the two is that `transform->minlen` calculates the exact padding length (which is not necessarily the block size: it's true in most cases, but not for mac-then-encrypt with a MAC size that isn't a multiple of the block size) whereas `get_record_expansion` uses the maximum padding size.

We could either have a function that takes a flag to indicate “padding for empty record” vs “maximum padding size”, or a pair of functions, one that returns the constant part of the overhead and one that returns the part that gets padded.

Reply:
@gilles-peskine-arm 

I suggest the following: We introduce a function `mbedtls_ssl_transform_get_expansion` which takes an SSL transform and an optional plaintext length, and returns pre-expansion and post-expansion values for the transform. If the plaintext length is specified, both must be accurate; if not, the postexpansion must be an overapproximation for all plaintext lengths. Pre-expansions never vary, so it is always accurate.

With that, `minlen` can be defined as `pre_exp(0) + post_exp(0)`, and `mbedtls_ssl_get_record_expansion()` can be defined as `mbedtls_ssl_hdr_len + pre_exp(unspecified) + post_exp(unspecified)`. 

Since `transform->minlen` is redundant, we might remove it from `mbedtls_ssl_transform` on that occasion.

The pre-expansion determines the offset of `ssl->out_msg` from `ssl->out_iv`, so we could also rewrite updating these pointers in terms of `mbedtls_ssl_transform_get_expansion`. 

The above proposal can not yet be implemented as stated since not all information determining a transform are actually contained in `mbedtls_ssl_transform`, for example the Encrypt-then-MAC flag. PR #1633 fixes this, so if you think the above sounds reasonable, I suggest we come back to it as soon as #1633 is merged, and stay with the current version of this PR until then.

WDYT?

Reply:
Ok, but why not add the missing information to the transform, or pass it separately?

Reply:
@gilles-peskine-arm I don't want to mix up this PR with #1633, which does add the missing fields.

Reply:
Ok, then let's stick with the duplicated calculation for now.

Reply:
@gilles-peskine-arm Do you approve, then?

Reply:
I'm ok to keep the duplicated calculation until #1633 is merged. But please make the calculation clearer as per https://github.com/ARMmbed/mbedtls/pull/1915/files#r207555016

Reply:
@gilles-peskine-arm If you insist I will change it, but if we leave a thorough cleanup for post #1633 anyway, would you accept leaving this PR as is for now? I think the documentation is quite clear on the source of the arithmetic, and the PR has already seen two approvals and @mpg's work on https://github.com/ARMmbed/mbedtls/pull/1939 builds on it.

Reply:
I'd really prefer the clear code. It doesn't conflict with any of the changes in #1939, does it?

Reply:
@gilles-peskine-arm Done.